gdx/src/com/badlogic/gdx/utils/ObjectMap.java

```diff
@@ -1,846 +1,782 @@
+	 *
+	 * @param key       a K key that will be checked for equality if a similar-seeming key is found
+	 * @param placement as calculated by {@link #place(Object)}, almost always with {@code place(key)}
+	 * @return the location in the key array of key, if found, or -1 if it was not found.
+	 */
+	int locateKey (final K key, final int placement) {
+		for (int i = placement; ; i = i + 1 & mask) {
+			// empty space is available
+			if (keyTable[i] == null) {
+				return -1;
+			}
+			if (key.equals(keyTable[i])) {
+				return i;
+			}
+		}
+	}
+
+	/**
+	 * Returns the old value associated with the specified key, or null.
+	 */
+	public V put (K key, V value) {
+		if (key == null)
+			throw new IllegalArgumentException("key cannot be null.");
+		K[] keyTable = this.keyTable;
+		V[] valueTable = this.valueTable;
+		int b = place(key);
+		int loc = locateKey(key, b);
+		// an identical key already exists
+		if (loc != -1) {
+			V tv = valueTable[loc];
+			valueTable[loc] = value;
+			return tv;
+		}
+		for (int i = b; ; i = (i + 1) & mask) {
+			// space is available so we insert and break (resize is later)
+			if (keyTable[i] == null) {
+				keyTable[i] = key;
+				valueTable[i] = value;
+				break;
+			}
+		}
+		if (++size >= threshold) {
+			resize(keyTable.length << 1);
+		}
+		return null;
+	}
+
+	//	public void putAll (ObjectMap<K, V> map) {
+//		ensureCapacity(map.size);
+//		for (Entry<K, V> entry : map)
+//			put(entry.key, entry.value);
+//	}
+	public void putAll (ObjectMap<K, V> map) {
+		ensureCapacity(map.size);
+		final K[] keyTable = map.keyTable;
+		final V[] valueTable = map.valueTable;
+		K k;
+		for (int i = 0, n = keyTable.length; i < n; i++) {
+			if ((k = keyTable[i]) != null)
+				put(k, valueTable[i]);
+		}
+	}
+
+	/**
+	 * Skips checks for existing keys.
+	 */
+	private void putResize (K key, V value) {
+		K[] keyTable = this.keyTable;
+		V[] valueTable = this.valueTable;
+		for (int i = place(key); ; i = (i + 1) & mask) {
+			// space is available so we insert and break (resize is later)
+			if (keyTable[i] == null) {
+				keyTable[i] = key;
+				valueTable[i] = value;
+				break;
+			}
+		}
+		if (++size >= threshold) {
+			resize(keyTable.length << 1);
+		}
+	}
+
+	/**
+	 * Returns the value for the specified key, or null if the key is not in the map.
+	 */
+	public V get (K key) {
+		final int loc = locateKey(key);
+		return loc == -1 ? null : valueTable[loc];
+	}
+
+	/**
+	 * Returns the value for the specified key, or the default value if the key is not in the map.
+	 */
+	public V get (K key, V defaultValue) {
+		final int loc = locateKey(key);
+		return loc == -1 ? defaultValue : valueTable[loc];
+	}
+
+	public V remove (K key) {
+		int loc = locateKey(key);
+		if (loc == -1) {
+			return null;
+		}
+		V oldValue = valueTable[loc];
+		int nl = (loc + 1 & mask);
+		while ((key = keyTable[nl]) != null && nl != place(key)) {
```

Comment:
I wonder about finding the indices of the elements to shift, then shifting them with arraycopy. Dealing with wrapping around is a little annoying, could use two loops. Maybe only worthwhile for large maps near the threshold. Would not want to hurt performance for smaller maps.

Reply:
It's an option, I guess. I think the bulk of the time spent here is used by finding those indices of what to shift, and the arraycopy would require some complex wraparound code (since it copies a linear span of memory, and when wrapping there are two spans) that would probably counteract any performance gain. If you want to write it and you think this might benefit, feel free. I don't feel terribly confident that I would write that wrapping code correctly the first time.

Reply:
One advantage to two loops (from `loc` to the end and from 0 to `loc`) is we don't need to apply the mask each iteration. Just a thought, but if we used two loops everywhere, then we wouldn't need the mask at all and growing the tables would not be limited to powers of two.

Reply:
To add another thought after all this time: the mask requires POT table sizes, making 1073741824 the largest possible table size. Doubling that gives -2147483648 and failure. Needing such a huge map is likely rare, so we'd not like to affect performance for much more common smaller maps, but maybe there's some special case trickery that would allow us to use a 2147483647 table size?

Reply:
I've thought about it a bit; there's actually a maximum array size in Java that's slightly less than 2147483647, depending on VM: https://stackoverflow.com/a/3039805/786740 It's a bit of a challenge to test, for me, because the responses mention needing 13GB of RAM available...